### PR TITLE
make float(NA) return nan

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/_libs/missing.pyx [hunk 1]
```diff
@@ -403,6 +404,9 @@ class NAType(C_NAType):
     def __repr__(self) -> str:
         return "<NA>"
 
+    def __float__(self):
+        return NAN
```
Comment (twoertwein) [inline pandas/_libs/missing.pyx:408]: Could probably also use `float("nan")` - not sure how numpy defines `np.nan`

Comment (MarcoGorelli) [reply pandas/_libs/missing.pyx]: hey thanks for weighing in - do mean to return `float('nan')` instead? sure, but why?

Comment (twoertwein) [reply pandas/_libs/missing.pyx]: would avoid an import but `float("nan") is np.nan` fails so probably better to stick with `np.nan`

**Comment by jbrockmendel** [comment]
Cc @jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
Thanks for the ping! 

I think we need to think through some actual use cases to see if returning NaN makes more sense or not (in the linked matplotlib PR, the issue is more that they got object dtype array, and so they (or in pandas) should ensure they convert to a float dtype, and then you don't run into a `float(pd.NA)` issue, since you already got floats)

I think the main reason to _not_ have this automatic conversion is that NaN and NA are inherently different (and they behave different in several operations, such as in boolean context, in comparisons, ...).

**Comment by MarcoGorelli** [comment]
Sure, the issue still arises when using `np.asanyarray(ser)` though, as then you can't pass a `na_value` and it'll fail on `pd.NA`:
```
In [5]: df = pd.DataFrame({'a': [1,2,3], 'b': [4,5,pd.NA]}, dtype='Int64')

In [6]: np.asanyarray(df['b'], dtype='float')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [6], line 1
----> 1 np.asanyarray(df['b'], dtype='float')

File ~/pandas-dev/pandas/core/series.py:902, in Series.__array__(self, dtype)
    855 def __array__(self, dtype: npt.DTypeLike | None = None) -> np.ndarray:
    856     """
    857     Return the values as a NumPy array.
    858 
   (...)
    900           dtype='datetime64[ns]')
    901     """
--> 902     return np.asarray(self._values, dtype)

File ~/pandas-dev/pandas/core/arrays/masked.py:496, in BaseMaskedArray.__array__(self, dtype)
    491 def __array__(self, dtype: NpDtype | None = None) -> np.ndarray:
    492     """
    493     the array interface, return my values
    494     We return an object array here to preserve our scalar values
    495     """
--> 496     return self.to_numpy(dtype=dtype)

File ~/pandas-dev/pandas/core/arrays/masked.py:418, in BaseMaskedArray.to_numpy(self, dtype, copy, na_value)
    412 if self._hasna:
    413     if (
    414         not is_object_dtype(dtype)
    415         and not is_string_dtype(dtype)
    416         and na_value is libmissing.NA
    417     ):
--> 418         raise ValueError(
    419             f"cannot convert to '{dtype}'-dtype NumPy array "
    420             "with missing values. Specify an appropriate 'na_value' "
    421             "for this dtype."
    422         )
    423     # don't pass copy to astype -> always need a copy since we are mutating
    424     data = self._data.astype(dtype)

ValueError: cannot convert to 'float64'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype.
```

**Comment by jorisvandenbossche** [comment]
(just to have it written down here as well)

> Sure, the issue still arises when using `np.asanyarray(ser)` though, as then you can't pass a `na_value` and it'll fail on `pd.NA`:

That's still a consequence of defaulting to object dtype (what is happening here is that the `df` gets converted to a single object dtype array, and then numpy tries to convert this to float, running into this issue). 

So also for this example we can handle this in `__array__` if we would like to avoid going through object dtype (raise error, directly coerce to float with NaNs, etc)

**Pull request closed without merging**